[mlir,python] Expose replaceAllUsesExcept to Python bindings

Wheest opened the pull request:
Problem originally described in [the forums here](https://discourse.llvm.org/t/mlir-python-expose-replaceallusesexcept/83068/1).

Using the MLIR Python bindings, the method [`replaceAllUsesWith`](https://mlir.llvm.org/doxygen/classmlir_1_1Value.html#ac56b0fdb6246bcf7fa1805ba0eb71aa2) for `Value` is exposed, e.g.,

```python
orig_value.replace_all_uses_with(
    new_value               
)
```

However, in my use-case I am separating a block into multiple blocks, so thus want to exclude certain Operations from having their Values replaced (since I want them to diverge).

Within Value, we have [`replaceAllUsesExcept`](https://mlir.llvm.org/doxygen/classmlir_1_1Value.html#a9ec8d5c61f8a6aada4062f609372cce4), where we can pass the Operations which should be skipped.

This is not currently exposed in the Python bindings: this PR fixes this.  Adds `replace_all_uses_except`, which works with individual Operations, and lists of Operations.

Wheest committed:
Expose replaceAllUsesExcept to Python bindings

makslevental commented on mlir/lib/CAPI/IR/IR.cpp:
```diff
@@ -1009,6 +1010,31 @@ void mlirValueReplaceAllUsesOfWith(MlirValue oldValue, MlirValue newValue) {
+                                          MlirValue newValue,
+                                          MlirOperation *exceptions,
+                                          intptr_t numExceptions) {
+  auto oldValueCpp = unwrap(oldValue);
+  auto newValueCpp = unwrap(newValue);
```
nit: please spell out auto (this is an LLVM wide code style thing)

Wheest committed:
Replace auto with explicit types

makslevental commented on mlir/lib/Bindings/Python/IRCore.cpp:
```diff
@@ -3718,6 +3724,32 @@ void mlir::python::populateIRCore(py::module &m) {
           kValueReplaceAllUsesWithDocstring)
+      .def(
+          "replace_all_uses_except",
+          [](PyValue &self, PyValue &with, py::object exceptions) {
+            MlirValue selfValue = self.get();
+            MlirValue withValue = with.get();
+
+            // Check if 'exceptions' is a list
+            if (py::isinstance<py::list>(exceptions)) {
+              // Convert Python list to a vector of MlirOperations
+              std::vector<MlirOperation> exceptionOps;
+              for (py::handle exception : exceptions) {
+                exceptionOps.push_back(exception.cast<PyOperation &>().get());
+              }
+              mlirValueReplaceAllUsesExceptWithSet(
+                  selfValue, withValue, exceptionOps.data(),
+                  static_cast<intptr_t>(exceptionOps.size()));
+            } else {
+              // Assume 'exceptions' is a single Operation
+              MlirOperation exceptedUser =
+                  exceptions.cast<PyOperation &>().get();
+              mlirValueReplaceAllUsesExceptWithSingle(selfValue, withValue,
+                                                      exceptedUser);
+            }
```
you can actually take advantage of cpp ADL here and have two different `.def("replace_all_uses_except", ...` that separately handle list/set or single `except` (see [overloaded methods](https://pybind11.readthedocs.io/en/stable/classes.html#overloaded-methods) but ignore `py::overload_cast` - that's only for actually overloaded methods on cpp classes)

Wheest replied:
Good suggestion, that would be clearer and avoids the branch.

I've updated the file with that change.  I've also replaced the `std::vector` with a `llvm::SmallVector` with default size 4.

Wheest committed:
Use ADL to split single op and list calls

makslevental commented on mlir/lib/Bindings/Python/IRCore.cpp:
```diff
@@ -3718,6 +3724,36 @@ void mlir::python::populateIRCore(py::module &m) {
           kValueReplaceAllUsesWithDocstring)
+      .def(
+          "replace_all_uses_except",
+          [](PyValue &self, PyValue &with, PyOperation &exception) {
```
sorry i didn't notice - you can just do here (and below)
```suggestion
          [](MlirValue self, MlirValue with, MlirOperation exception) {
```

all of this magic over here in [PybindAdaptors.h](https://github.com/llvm/llvm-project/blob/825d9f6e0cea21c50a767e80be5ba80479d65f0e/mlir/include/mlir/Bindings/Python/PybindAdaptors.h#L66-L202) will take care of it for you.

Wheest replied:
I note that some existing code such as `replace_all_uses_with` also uses `PyValue &self, PyValue &with`, should I look at making that consistent in a future PR?

makslevental replied:
I took a quick look in `IRCore.cpp` before making the suggestion it seems those uses are because the various fields/methods of those classes are in fact exercised. Here, where you're just immediately calling the C API, you don't need any of that functionality. I mean ultimately this is a nit - it's up to you - you can save a couple of lines but it makes no difference (the magic in `PybindAdaptors.h` does exactly what you're doing here - `.get()`).

makslevental reviewed (approved):
LGTM! Thanks for the feature. Let me know when you're ready for me to merge.

Wheest committed:
Leverage PybindAdaptors and fix test

Wheest committed:
Revert MlirOperation PybindAdaptor for ADL

ftynse commented on mlir/lib/Bindings/Python/IRCore.cpp:
```diff
@@ -3718,6 +3724,36 @@ void mlir::python::populateIRCore(py::module &m) {
           kValueReplaceAllUsesWithDocstring)
+      .def(
+          "replace_all_uses_except",
+          [](PyValue &self, PyValue &with, PyOperation &exception) {
```
It should be possible to take `MlirValue` here directly and avoid the conversion below.

ftynse commented on mlir/lib/Bindings/Python/IRCore.cpp:
```diff
@@ -3718,6 +3724,29 @@ void mlir::python::populateIRCore(py::module &m) {
+          "replace_all_uses_except",
+          [](MlirValue self, MlirValue with, py::list exceptions) {
+            // Convert Python list to a SmallVector of MlirOperations
+            llvm::SmallVector<MlirOperation, 4> exceptionOps;
```
```suggestion
            llvm::SmallVector<MlirOperation> exceptionOps;
```

Don't specify the number of stack elements unless there is a strong reason to pick one.

Wheest committed:
Remove extraneous overloads

Wheest commented on mlir/include/mlir-c/IR.h:
```diff
@@ -956,6 +956,22 @@ MLIR_CAPI_EXPORTED MlirOpOperand mlirValueGetFirstUse(MlirValue value);
+/// 'exceptedUser'.
+MLIR_CAPI_EXPORTED void
+mlirValueReplaceAllUsesExceptWithSingle(MlirValue of, MlirValue with,
+                                        MlirOperation exceptedUser);
```
Good point, the extra code footprint isn't worth it.  Have unified.

Wheest commented on mlir/include/mlir-c/IR.h:
```diff
@@ -956,6 +956,22 @@ MLIR_CAPI_EXPORTED MlirOpOperand mlirValueGetFirstUse(MlirValue value);
+MLIR_CAPI_EXPORTED void
+mlirValueReplaceAllUsesExceptWithSet(MlirValue of, MlirValue with,
+                                     MlirOperation *exceptions,
+                                     intptr_t numExceptions);
```
Will bear that in mind, is this in a formal style guide, or a convention that is used across the project but not explicitly documented?

Wheest commented on mlir/include/mlir-c/IR.h:
```diff
@@ -956,6 +956,22 @@ MLIR_CAPI_EXPORTED MlirOpOperand mlirValueGetFirstUse(MlirValue value);
+/// 'exceptions'. The 'exceptions' parameter is an array of MlirOperation
+/// pointers with a length of 'numExceptions'.
+MLIR_CAPI_EXPORTED void
+mlirValueReplaceAllUsesExceptWithSet(MlirValue of, MlirValue with,
```
Have done this, and also unified into a single function (no more `mlirValueReplaceAllUsesExceptWithSet` + `mlirValueReplaceAllUsesExceptWithSingle`.

Wheest commented:
Cheers both!  I'm happy to get it merged now if there are no further comments.

I've implemented @ftynse's suggestions.

For @makslevental, I added the PybindAdaptor, stuff, worth it I reckon.

Only things to point out are:

1. If you note `d9c3cd7`, I had to revert `MlirOperation exception` back to `PyOperation &exception`, as this was not compatible with the ADL:

Calling `value.replace_all_uses_except(value2, [op1, op2])` raised the error

```
# | TypeError: Expected an MLIR object (got [<mlir._mlir_libs._mlir.ir.Operation object at 0x71b9f1341270>, <mlir._mlir_libs._mlir.ir.Operation object at 0x71b9f1342730>]).
```

2. I fixed the first test `testValueReplaceAllUsesWithExcept`

```diff
- value.replace_all_uses_except(value2, [op1])
+ value.replace_all_uses_except(value2, op1)
```

As you can see, before I was testing passing a list of an Operation, which doesn't cover the behaviour I wanted to check.

Btw, thanks for running https://github.com/makslevental/mlir-wheels/, it has been a boon to my hacking!

makslevental commented:
cool - @ftynse will probably approve soon :smiley:
